### PR TITLE
[FAT-9774]-Cancellation Flow

### PR DESCRIPTION
**Pull request by SinghAdes**
## Purpose
_Describe the purpose of the pull request._

## Approach
_How does this change fulfill the purpose?_

### TODOS and Open Questions
- [ ] Use GitHub checklists. When solved, check the box and explain the answer.

### Learning
_Describe the research stage. Add links to blog posts, patterns, libraries or addons used to solve this problem._

### CODE REVIEW

**Comment by SinghAdes** [comment]
image

**Review by pavankumar181 — APPROVED** [review]

**Review by gurleenkaurbp — APPROVED** [review]

File: mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-lending-flow.feature [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+  Background:
+    * url baseUrl
+
+    * callonce login testAdmin
+    * def okapitokenUser = okapitoken
+    * def headersUser = { 'Content-Type': 'application/json', 'x-okapi-token': '#(okapitokenUser)', 'Accept': 'application/json, text/plain'  }
+    * configure headers = headersUser
+    * callonce variables
+
+  Scenario: Cancel DCB Transaction manually
```
Comment (Vignesh-kalyanasundaram) [inline mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-lending-flow.feature:12]: Please add one more scenario that create a transaction, cancel the request by directly calling request API(Mocking UI action), then verify the status of the transaction. It should be cancelled.

File: mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-lending-flow.feature [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    And match $.status == 'ITEM_CHECKED_OUT'
+    And match $.role == 'LENDER'
+
+    Given path 'transactions' , transactionId , 'status'
+    And request updateToCancelRequest
+    When method PUT
+    Then status 400
+    And match $.errors[0].message == 'Cannot cancel transaction dcbTransactionId: 040. Transaction already in status: ITEM_CHECKED_OUT: '
+
+  Scenario: Cancel DCB Transaction manually after ITEM_CHECKED_IN
```
Comment (Vignesh-kalyanasundaram) [inline mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-lending-flow.feature:189]: I think after checked_in, cancellation should not be allowed

File: mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-flow-for-borrowing-pickup-flow.feature [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+  Background:
+    * url baseUrl
+
+    * callonce login testAdmin
+    * def okapitokenUser = okapitoken
+    * def headersUser = { 'Content-Type': 'application/json', 'x-okapi-token': '#(okapitokenUser)', 'Accept': 'application/json, text/plain'  }
+    * configure headers = headersUser
+    * callonce variables
+
+  Scenario: Cancel DCB Transaction manually
```
Comment (Vignesh-kalyanasundaram) [inline mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-flow-for-borrowing-pickup-flow.feature:12]: Please add one more scenario, try to cancel dcb transaction which is in closed status. This should give 4xx error

File: mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-flow-for-pickup-flow.feature [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+  Background:
+    * url baseUrl
+
+    * callonce login testAdmin
+    * def okapitokenUser = okapitoken
+    * def headersUser = { 'Content-Type': 'application/json', 'x-okapi-token': '#(okapitokenUser)', 'Accept': 'application/json, text/plain'  }
+    * configure headers = headersUser
+    * callonce variables
+
+  Scenario: Cancel DCB Transaction manually
```
Comment (Vignesh-kalyanasundaram) [inline mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-flow-for-pickup-flow.feature:12]: Please add one more scenario, try to cancel dcb transaction which is in closed status. This should give 4xx error

File: mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-flow-for-borrower-flow.feature [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    And match $.status == 'ITEM_CHECKED_OUT'
+    And match $.role == 'BORROWER'
+
+    Given path 'transactions' , transactionId , 'status'
+    And request updateToCancelRequest
+    When method PUT
+    Then status 400
+    And match $.errors[0].message == 'Cannot cancel transaction dcbTransactionId: 400. Transaction already in status: ITEM_CHECKED_OUT: '
+
+   Scenario: Cancel DCB Transaction manually after ITEM_CHECKED_IN
```
Comment (Vignesh-kalyanasundaram) [inline mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-flow-for-borrower-flow.feature:150]: I think we can merge all the negative scenarios into a single one. Like below.
Once the transaction status is Item_checked_out, we will try to cancel the dcb transaction. We will get 4xx error. Now still the transaction status is ITEM_CHECKED_OUT, so we can use this transaction to updated the status to ITEM_CHECKED_IN and then we will try to cancel the dcb transaction. By this way, we dont need to create multiple scenarios for cancellation negative scenarios.

File: mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-lending-flow.feature (hunk 1 above)
Comment (SinghAdes) [reply mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-lending-flow.feature:12]: fixed

File: mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-lending-flow.feature (hunk 2 above)
Comment (SinghAdes) [reply mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-lending-flow.feature:208]: fixed

File: mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-flow-for-borrowing-pickup-flow.feature (hunk 3 above)
Comment (SinghAdes) [reply mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-flow-for-borrowing-pickup-flow.feature:12]: fixed

File: mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-flow-for-pickup-flow.feature (hunk 4 above)
Comment (SinghAdes) [reply mod-dcb/src/main/resources/volaris/mod-dcb/features/cancelling-flow-for-pickup-flow.feature:12]: fixed

**Pull request merged**